### PR TITLE
Implement a way for ParameterHandler to alias parameters.

### PR DESCRIPTION
**Pull request by bangerth**
This also allows for deprecation notices to be printed to the screen.
There should be tests for basically everything except for the
deprecation message to the screen for which I couldn't come up with a
way to get it into the test output. I did verify it by hand, though.

### CODE REVIEW

**Comment by bangerth** [comment]
Addressed part of #659

**Comment by tamiko** [comment]
Looks good. Please rebase and merge.

File: include/deal.II/base/parameter_handler.h [hunk 1]
```diff
@@ -1648,7 +1648,56 @@ class ParameterHandler : public Subscriptor
...
+   * The primary purpose of this function is to allow for a backward
+   * compatible way of changing names in input files of applications
+   * for which backward compatibility is important. This can be
+   * achieved by changing the name of the parameter in the call to
+   * declare_entry(), and then creating an alias that maps the old
+   * name to the new name. This way, old input files can continue
+   * to refer to parameters under the old name, and they will
+   * automatically be mapped to the new parameter name.
+   *
+   * It is valid to reference in a parameter file to the same parameter
```
Comment (tjhei) [inline include/deal.II/base/parameter_handler.h]: weird sentence, can you rephrase?

File: include/deal.II/base/parameter_handler.h [hunk 2]
```diff
@@ -1648,7 +1648,56 @@ class ParameterHandler : public Subscriptor
...
+   * multiple times. The value that will ultimately be chosen in such
+   * cases is simply the last value set. This rule also applies to
```
Comment (tjhei) [inline include/deal.II/base/parameter_handler.h]: same here, remove "is"?

File: source/base/parameter_handler.cc [hunk 3]
```diff
@@ -1610,6 +1634,55 @@ ParameterHandler::declare_entry (const std::string           &entry,
...
+          ExcMessage ("You are trying to declare an alias entry <"
+                      + alias_name +
+                      "> that references an entry <"
+                      + existing_entry_name +
+                      ">, but the latter does not exist."));
+
+  // now also make sure that if the alias has already been
+  // declared, that it is also an alias and refers to the same
+  // entry
+  if (entries->get_optional<std::string>(get_current_full_path(alias_name)))
```
Comment (tjhei) [inline source/base/parameter_handler.cc]: I think a call to ```declare_alias``` should fail if ```alias_name``` already exists. What is the reason for allowing it as long as it points to the same parameter?

**Comment by bangerth** [comment]
@tjhei : I didn't see your mails until after merging as suggested by @tamiko . Will take care of your comments in a separate pull request -- keep sending suggestions.

**Comment by tjhei** [comment]
What happens right now if you alias an alias, does that work (and should it?)?

**Comment by tjhei** [comment]
(looks good otherwise)

File: source/base/parameter_handler.cc (hunk 3 above)
Comment (bangerth) [inline source/base/parameter_handler.cc]: Dunno. We allow declaring the same parameter entry multiple times. I thought I'd just mirror existing behavior. Feel strongly about this?

**Comment by bangerth** [comment]
Indeed, there was an assertion missing to ensure that you're not aliasing an existing alias (or a section).

**Comment by bangerth** [comment]
@tjhei  I've addressed your additional comments in #689

File: source/base/parameter_handler.cc (hunk 3 above)
Comment (tjhei) [inline source/base/parameter_handler.cc]: no, just wondering because it would have been easier/shorter to check here. :-)
